### DISCUSSION TITLE
@TransactionalEventListener와 @EventListener의 차이점 및 @Transactional의 다른 용도

### DISCUSSION THREAD

**Discussion by soeunnPark** [Kernel360/F2_PLAY_BADMINTON_BE#520] [General]
이미지 업로드를 비동기적으로 처리하게 되면서, Event Publisher를 사용하게 되었습니다. 이때 TransactionalEventListener와 EventListener 중 어느 것을 사용할지 아래와 같은 의견을 주고받게 되었습니다!

![image](https://github.com/user-attachments/assets/5aeebbdf-22f3-4fa8-a4e3-5df31dfc15f4)

`@Transactional` 어노테이션을 붙이면 RuntimeException이 발생했을 때도 롤백시킵니다. 따라서 지금 `@Transactional`로 묶어 놓은 작업에 DB로 읽기/쓰기 작업은 없지만, 하나의 작업으로 묶은 다음, `@TransactionalEventListener`를 통해 1, 2, 4에서 예외 발생 시 이벤트가 처리되지 않는 것을 보장하고 있습니다. 

아래의 경우를 논의하고 있습니다. 

1. publishEvent() 메서드를 제일 아래쪽에 쓰면 해결됨
💁‍♂️: 이벤트를 발행하는 코드를 제일 아래쪽에 위치하면, 윗줄에서 예외 발생 시 이벤트를 처리하지 않는다.
💁: 코드의 순서로 이를 보장하기 보다는, `@TransactionalEventListener`로 확실하게 보장하는 것이 좋을 것 같습니다.
2. DB 작업이 없는데 `@Transactional`을 쓰는게 맞아?
💁‍♂️: DB에 읽기/쓰기를 하지 않는데 `@Transactional`을 써야 하나요?
💁: `@Transactional`로 묶여 있는 스레드에서 예외가 발생하면 실패로 간주됩니다. 만약 이미지 업로드를 하나의 작업 단위로 묶고, @TransactionalEventListener를 통해 모든 작업이 성공했을 때에만 이벤트를 처리하고 싶다면, `@Transactional`을 사용해도 괜찮지 않을까요?

추가적으로, 이미지 업로드를 Event Handler로 비동기적으로 처리하는 것이 적절할까요? 아래의 문제가 있을 것 같습니다.

이미지 업로드를 비동기적으로 처리하게 되면, 이벤트 리스너에서 예외가 발생하더라도, API는 정상적으로 이미지 URL 을 반환합니다. 즉, 이미지가 S3에 업로드되지 않았는데도, 사용자는 정상적으로 올라갔구나~ 생각하게 됩니다.

다만, 이미지 업로드를 동기적으로 처리하게 되면, 시간이 오래 걸립니다. 현재 이미지 최적화 작업(Webp로 변환)을 하고 있어 5초 정도 걸립니다.

이러한 문제를 어떻게 해결할 수 있을까요?

**Comment by soeunnPark** [comment]
아래는 대면으로 대화한 내용입니다. 

1. DB 읽기/쓰기가 없다면 `@Transactional`을 사용하지 않는다. 다만 saveImage 외부에 `@Transactional` 이 걸려있다. 그렇기에 `@TransactionalEventListener`를 사용하는 것이 자연스럽다.
2. 무조건 `@TransactionalEventListener`를 쓰는 것이 좋은게 아니다. 요구사항에 따라 달라진다. 만약, 동호회 생성이 실패하던, 성공하던 메일이 전송되어야 한다면 `@EventListener`를 사용하는 것이다. 지금의 경우 트랜잭션이 정상적으로 종료된 이후에 이메일을 발송해야 하기 때문에 `@TransactionalEventListener`가 적합하다.
3. transactionId, status를 활용할 수 있다. 파일 업로드 화면에 접속하면, 업로드 용으로 아이디를 하나 발급한 다음, 이를 고유키로 활용한다. 파일을 업로드하면 S3에 업로드를 하고, PENDING 상태로 DB에 이미지 url과 함께 저장한다. 이후 최종 등록 버튼을 누를 때 해당하는 transactionId의 상태를 COMPLETED 상태로 변경한다. PENDING 상태로 S3에 업로드가 되어 있는 이미지는 배치를 2시간 간격으로 돌려 삭제한다.  
5. 비동기적으로 처리한다면, 프론트에서 추가적인 처리가 필요하다. 그래서 시간이 1~2초 더 걸리더라도 동기적으로 처리하는게 더 간단할 수 있다. 
6. 기존처럼 비동기 처리 방식을 유지하려면, presizeUrl로 접근했을 때 이미지 성공 / 실패 여부에 따른 retry 로직이 추가되어야 한다. 프론트에서 이미지가 로딩되지 않는다면, 1초 뒤에 다시 해당 S3 Url 에 대해 요청을 해서 다시 처리해야 한다. 
7. Apache Tika (파일 유효성 검사 할 때 사용하는 라이브러리) 를 알아보자.

**Comment by km2535** [comment]
그럼 이렇게 수정된 코드에서는 Transactional이 필요할까요?
```java

	@Transactional
	public String saveImage(ImageUploadRequest request) {
		try {
			byte[] byteFile = request.multipartFile().getBytes();

			String originalNam = request.multipartFile().getOriginalFilename();
			// uuid 객체 생성
			String uuid = UUID.randomUUID().toString();
			String extension = getFileExtension(Objects.requireNonNull(originalNam));

			// 비동기 처리
			eventPublisher.publishEvent(new ClubImageEvent(byteFile, originalNam, uuid));

			// url 만드는 곳으로 전달
			return preReturnUrl(uuid, extension);
		} catch (IOException e) {
			throw new EmptyFileException(e);
		}
```

**Reply by soeunnPark** [reply]
네! 부모 트랜잭션이 있다면, 부모 트랜잭션이 합쳐질 것이고
부모 트랜잭션이 없다면 자신이 트랜잭션을 새로 생성할 거라 괜찮을 것 같습니다.
